op_crates/crypto/lib.rs

```diff
@@ -43,6 +81,381 @@ pub fn op_crypto_get_random_values(
+      );
+
+      // Determine agreement from algorithm named_curve.
+      let agreement: &RingAlgorithm = args
+        .algorithm
+        .named_curve
+        .ok_or_else(|| {
+          WebCryptoError::MissingArgument("namedCurve".to_string())
+        })?
+        .try_into()?;
+      // Generate private key from agreement and ring rng.
+      let rng = RingRand::SystemRandom::new();
+      let private_key = EphemeralPrivateKey::generate(&agreement, &rng)?;
+      // Extract public key.
+      let public_key = private_key.compute_public_key()?;
+      // Create webcrypto keypair.
+      let webcrypto_key_public =
+        WebCryptoKey::new_public(algorithm, extractable, vec![]);
+      let webcrypto_key_private =
+        WebCryptoKey::new_private(algorithm, extractable, vec![]);
+      let crypto_key = WebCryptoKeyPair::new(
+        webcrypto_key_public.clone(),
+        webcrypto_key_private.clone(),
+      );
+
+      JSCryptoKey::Pair {
+        key: crypto_key,
+        private_rid: state.resource_table.add(CryptoKeyResource {
+          crypto_key: webcrypto_key_private,
+          key: private_key,
+          hash: args.algorithm.hash,
+        }),
+        public_rid: state.resource_table.add(CryptoKeyResource {
+          crypto_key: webcrypto_key_public,
+          key: public_key,
+          hash: args.algorithm.hash,
+        }),
+      }
+    }
+    Algorithm::Ecdsa => {
+      validate_usage!(args.key_usages, vec![KeyUsage::Sign, KeyUsage::Verify]);
+
+      let curve: &EcdsaSigningAlgorithm = args
+        .algorithm
+        .named_curve
+        .ok_or_else(|| {
+          WebCryptoError::MissingArgument("namedCurve".to_string())
+        })?
+        .try_into()?;
+
+      let rng = RingRand::SystemRandom::new();
+      let pkcs8 = EcdsaKeyPair::generate_pkcs8(curve, &rng)?;
+      let private_key = EcdsaKeyPair::from_pkcs8(&curve, pkcs8.as_ref())?;
+      // Create webcrypto keypair.
+      let webcrypto_key_public =
+        WebCryptoKey::new_public(algorithm, extractable, vec![]);
+      let crypto_key =
+        WebCryptoKey::new_private(algorithm, extractable, vec![]);
+
+      let resource = CryptoKeyResource {
+        crypto_key: crypto_key.clone(),
+        key: private_key,
+        hash: args.algorithm.hash,
+      };
+
+      JSCryptoKey::Single {
+        key: crypto_key,
+        rid: state.resource_table.add(resource),
+      }
+    }
+    Algorithm::Hmac => {
+      validate_usage!(args.key_usages, vec![KeyUsage::Sign, KeyUsage::Verify]);
+
+      let mut hash: HmacAlgorithm = args
+        .algorithm
+        .hash
+        .ok_or_else(|| WebCryptoError::MissingArgument("hash".to_string()))?
+        .into();
+      let rng = RingRand::SystemRandom::new();
+      
+      // if let Some(length) = args.algorithm.length {
+      //   let mut alg = hash.digest_algorithm();
+      //   alg.output_len = length as usize;
+      //   alg.chaining_len = length as usize;
+      // };
```

Comment:
`ring` expects digest algorithm at key generation thus we cannot modify the digest output length (thanks to ring's easy-to-misuse api).

Currently `generateKey` uses:
P256 -> SHA256
P384 -> SHA384
these are the recommended signing algorithms as per the [docs](https://briansmith.org/rustdoc/ring/signature/index.html#statics).
(also idk why people would use a custom length :thinking: (maybe just the spec))

Reply:
Upstream issue: https://github.com/briansmith/ring/issues/1216